src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.SparkSession
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.index.IndexConstants.{OPTIMIZE_MODE_FULL, OPTIMIZE_MODE_QUICK}
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, OptimizeActionEvent}
+import com.microsoft.hyperspace.util.{HyperspaceConf, PathUtils}
+
+/**
+ * Optimize Action provides an optimize support for indexes where small index files
+ * can be merged into larger ones for better index performance.
+ *
+ * Algorithm outline:
+ * 1. Collect all the currently valid index files.
+ * 2. Split files into small and large, based on a threshold.
+ * 3. Bucketwise combine smaller files into 1 file per bucket.
+ * 4. Update index snapshot to remove small files and keep large files + newly created files.
+ *
+ * NOTE: This is an index-only operation. It does not look at the current state of
+ * the data at all. If the data was changed after index creation, optimize will NOT
+ * include the data changes.
+ *
+ * Available modes:
+ * `Quick` mode: This mode allows for fast optimization. Files smaller than a
+ * predefined threshold "spark.hyperspace.index.optimize.threshold" will be picked
+ * for compaction.
+ *
+ * `Full` mode: This allows for slow but complete optimization. ALL files are
+ * eligible for compaction. Compaction threshold is inifinity.
+ *
+ * TODO: Optimize can be a no-op if there is only one small file per bucket.
+ *  https://github.com/microsoft/hyperspace/issues/204.
+ *
+ * @param spark SparkSession
+ * @param logManager IndexLogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ * @param mode Acceptable optimize modes are `quick` and `full`.
+ */
+class OptimizeAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager,
+    mode: String)
+    extends RefreshActionBase(spark, logManager, dataManager) {
```

Comment:
What are the things you are using from `RefreshActionBase`?

Reply:
Oh i used it for `previousIndexLogEntry` but it doesn't look right. I will update it to CreateActionBase

Reply:
thanks @imback82 , modified the hierarchy to extend from CreateActionBase. It uses `indexDataPath`. We can also move that function directly here and not depend on any other class. Please let me know